### PR TITLE
Update dependencies should never run concurrently

### PR DESCRIPTION
**Pull request by arichiardi**
Here is the fix, with the tests is more consistent, but i was able to just tear the commit to pieces :wink:

```
feat(leiningen): isolate update dependencies execution and added cancelability

The patch introduces two features. The first disallow the possibility to run concurrent dependency
updates. An atom has been added in ccw.leiningen.classpath-container in order to control this.
The second is WorkspaceJob cancel support: the implementation can be polished with macros but at the
moment functions are passed around (especially in get-project-dependencies) in order to throw
OperationCancelledException in the appropriate place and a new cancelable-workspace-job has been
added in ccw.eclipse.

The cancel response is not immediate as only get-project-dependencies has been harnessed with
breaking capabilities.
```

### CODE REVIEW

**Comment by arichiardi** [comment]
It failes because of an addition ```SmokeTests``` I did in the other PR

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 1]
```diff
@@ -238,6 +240,35 @@
...
 
+(defn- resource-identifier
+  "From Eclipse's doc: Resources are identified by type and by their path, which is similar to a file system path."
+  [r]
+  (into [] ((juxt #(.getType %1) #(.getFullPath %1)) r)))
+
+(defprotocol GenerateIdentifier
+  (identifier [this] "Coerce this to an immutable unique identifier according to its type"))
+
+(extend-protocol GenerateIdentifier
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: what's the purpose of this GenerateIdentifier protocol?

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 2]
```diff
@@ -620,10 +651,58 @@
...
         (catch Exception e
           (CCWPlugin/createErrorStatus (format "Unexpected exception while executing Job %s" name), e))))))
 
+(defn cancelable-workspace-job
+  "Create a WorkspaceJob with name, error-handler, cancel-handler and finally-handler.
+  Delegate runInWorkspace to (f monitor).
+  Takes care of returning Status/OK_STATUS, STATUS/CANCEL_STATUS or error.
+  The handlers finally-handler! and cancel-handler! should accept the
+  WorkspaceJob as parameter while error-handler should accept the
+  WorkspaceJob and the Exception (in a vector).
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: why in a vector?

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj [hunk 3]
```diff
@@ -40,6 +41,13 @@
 (def LEININGEN_CLASSPATH_CONTAINER_PROBLEM_MARKER_TYPE
   "ccw.leiningen.problemmarkers.classpathcontainer")
 
+(defonce ^{:private true
+           :author "Andrea Richiardi"
+           :doc "Atom containing the state of this namespace.
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Red lights flashing, red lights flashing when talking about the "state of a namespace" as if a namespace was a class ...

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj [hunk 4]
```diff
@@ -162,19 +170,22 @@
...
-      (filter #(re-find #"\.(jar|zip)$" (.getName ^File %)))
-      (sort-by #(.getName ^File %))
-      (map #(ser-dep %
-                     default-native-platform-path
-                     (get srcmap %)
-                     #_(or #_(u/lein-native-dependency-path lein-project %) ;; TODO make this work :-(
+    (for [dep (->> dependencies
+                   (filter #(re-find #"\.(jar|zip)$" (.getName ^File %)))
+                   (sort-by #(.getName ^File %)))]
+      (do (break!)
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: `for` is not an imperative for-loop. It will return a lazy sequence. So you don't know when the call to `break!` will occur, unless you explicitly `(dorun)` over the sequence. but here, then, it would be more appropriate to directly use `doseq`, I guess.

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj [hunk 5]
```diff
@@ -353,50 +364,102 @@
...
+                     ;; this rule is OK because we know the job needs it and will not take too long
+                     (.setRule (e/workspace-root))
+                     (.schedule))))
+               (if-not (.isCanceled monitor)
+                 Status/OK_STATUS
+                 Status/CANCEL_STATUS))
+             :error-handler! error-fn
+             :finally-handler! finally-fn)
+        new-state (swap! state start-updating project-id job)]
+    (when (can-update? new-state project-id)
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: this code is not free from race conditions

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj [hunk 6]
```diff
@@ -353,50 +364,102 @@
...
+                              (fn [^IProgressMonitor monitor]
+                                (refresh-target-folder java-project monitor)
+                                Status/OK_STATUS))
+                           (.setUser true)
+                           (.schedule)))
+        job (e/cancelable-workspace-job
+             (format "Update project dependencies for project %s" project-name)
+             (fn [^IProgressMonitor monitor]
+               (letfn [(cancel! [] (e/throw-op-canceled! (format "The update of %s was canceled" project-name) monitor))]
+                 (let [_ (cancel!)
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: How does `cancel!` work? I think I've not understood yet

File: ccw.core/src/clj/ccw/eclipse.clj (hunk 2 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/eclipse.clj]: to take advantage of identity as default :smile:

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 3 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Alarm alarm! Right (and thanks)...I am open to suggestions :wink:

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 4 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Oh gasp, wow, great point!

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 7]
```diff
@@ -620,10 +651,58 @@
...
+
+  Note1: f should always check for (.isCanceled monitor) and break the
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: instead of mentioning "check for (.isCanceled monitor)", why not mention the name of the function that wraps this call in `ccw.eclipse`?

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 8]
```diff
@@ -620,10 +651,58 @@
...
+  computation (throwing either InterruptedException or
+  OperationCanceledException) in order to react to user's cancel action.
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: what's the difference between InterruptedException and OperationCanceledException?

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 5 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Mmm, I thought a lot about it, you mean the swap part?
It should be, as ```can-update?``` (in the map) is written only and visible only by one of the concurrent threads that are calling ```swap!```, whoever wins.
But I can be bugged in my head, could you provide a counter-example?

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 9]
```diff
@@ -620,10 +651,58 @@
...
+  Note2: error-handler! should accept the exception and if returns
+  something, it will interpreted as the message (string) together with
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: it will `be` interpreted

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 6 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Basically it should cancel the execution when appropriate.
Now it throws when monitor's ```isCanceled``` is true.

File: ccw.core/src/clj/ccw/eclipse.clj (hunk 8 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/eclipse.clj]: The former is the mechanism used to stop a ```java.lang.Thread```, the latter is Eclipse's specific.
I handled both cases as one would also want to interrupt the ```currentThread```, when trying to perform a cancel.

File: ccw.core/src/clj/ccw/eclipse.clj (hunk 9 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/eclipse.clj]: tnx!

File: ccw.core/src/clj/ccw/eclipse.clj [hunk 10]
```diff
@@ -620,10 +651,58 @@
         (let [s (f monitor)]
           (if (instance? IStatus s)
             s
-            (Status/OK_STATUS)))
+            Status/OK_STATUS))
         (catch Exception e
           (CCWPlugin/createErrorStatus (format "Unexpected exception while executing Job %s" name), e))))))
 
+(defn cancelable-workspace-job
```
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/eclipse.clj]: why not just enhance the existing workspace-job function?

Comment (arichiardi) [inline ccw.core/src/clj/ccw/eclipse.clj]: I wanted to isolate such a big change...when the test will confirm that everything is fine...well I can just change the name (I know I know, overly defensive programming...)

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 5 above)
Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: I don't think we should just be happy with "whoever wins". Unless there's a bug, if `update-dependencies` is called twice, it's because there has been changes to be taken into account. Meaning that the current job should stop as far as possible, to let the second job work with (presumably) more recent data.

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: I interpreted it as a guard for concurrent executions, not actually knowing the algorithm behind (sorry, I am still a ```ccw``` dev newbie :smile:)...in any case, now it should be easy to employ a "last wins" policy.

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: you may be interested in reading this http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fruntime_jobs_rules.htm
and this https://wiki.eclipse.org/FAQ_How_do_I_prevent_two_jobs_from_running_at_the_same_time%3F

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Also, as we can see, even a tiny problem can cause lots of review / dialog. I would isolate the part dealing with user cancelation from the part dealing with concurrent update dependencies job.

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Wow, way easier! But my solution keeps the jobs instance so that we can cancel them on shutdown and it felt closer to my Clojure feelings...
And again, it makes sense now to implement a "last wins" policy, does it?

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: is the "cancel on shutdown" policy implemented in the PR? I may have overlooked it.

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Af for keeping the jobs instance, I guess a more general solution would be to have a group for all the CCW jobs, or another way to keep all jobs under a tag, and then query the job manager. No need for us to manage a separate registry.

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: See this, for instance: http://help.eclipse.org/juno/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcore%2Fruntime%2Fjobs%2FIJobManager.html

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Laurent, I didn't understand what you meant with review/dialog? Yes the canceling was included here as it just adds without mingling with the other behaviour.

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: sorry it's not a group, it's rather called a `family` as far as I understand Eclipse Jobs

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: I read the links, and yes, it might be better to use Eclipse's facilities.

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: What bothers me a little, is that I don't know if we are creating a general mechanism as a way to work around a bug (or a category of bugs) instead of studying why there are two jobs, and whether it's appropriate or not, and what we should do. 




I'd like to take a step back and reproduce and analyze what's happening here. Will do this tonight or tomorrow. 



—
Envoyé avec Mailbox

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Super fair enough, I did not have the patience and time to go over all the calls...I have noticed that ```update-project-dependencies``` is called twice, in two different points in the code...and on project creation this is evident. Moreover, I wanted to give a spin to cancelling, as sometimes you really want to stop crunching dependencies (as in my case, when I had slow internet connection).

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: I understand for cancelling. But it is definitely a separate concern and should be addressed in another PR.

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Ok, if you don't mind, I will revert this branch (keeping a copy) to ```master``` regarding ```classpath_container``` and keep the cancelling part. I had some macro in mind to ease the pain of calling ```cancel!``` every time.
In any case it will be priority number two, I still need to improve the test and fix (with test) the ```Namespace``` bug.

Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: Mm...If you make progress it is better for you to open a new PR on a brand new branch I guees...therefore it is probably better to close this...

Comment (laurentpetit) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: I may find a better solution, or not. We'll see :-)

File: ccw.core/src/clj/ccw/eclipse.clj (hunk 1 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/eclipse.clj]: It was to get a unique identifier based on the type...

File: ccw.core/src/clj/ccw/leiningen/classpath_container.clj (hunk 5 above)
Comment (arichiardi) [inline ccw.core/src/clj/ccw/leiningen/classpath_container.clj]: For sure! :+1:

**Comment by laurentpetit** [comment]
OK, so I managed to identify the code that was doing things unnecessarily.
I think I have tested all the other paths that could have been affected by my change, and that I've not introduced a regression. The fix is in commit https://github.com/laurentpetit/ccw/commit/b5647322a09181d01ddfc90c1d7e14ad8ad452cc

**Comment by laurentpetit** [comment]
I will close this PR, then.

**Comment by arichiardi** [comment]
:+1:

**Comment by arichiardi** [comment]
By the way, if you want I can test some behavior (like update dependencies)...something that you've noticed should happen on the ui...for example I have noticed a lot of nested WorkspaceJob and wondered if I should see them all as dialogs..
